### ISSUE TITLE
Optional is not narrowed when using structural matching

### ISSUE THREAD

**Issue by scullionw** [microsoft/pyright#3019]
**Describe the bug**
Pyright doesn't understand the new structural matching syntax in the case of matching on Optional values

**To Reproduce**
Have the following code in a python file in vscode

**Expected behavior**
Pyright should understand that `None` has already been matched and narrow the type to `int`

**Screenshots or Code**
```python
def some_or_none() -> int | None:
    return 5

def take(n: int):
    pass

match some_or_none():
    case None:
        print("we have none!")
    case val:
        take(val)
```
<img width="1042" alt="Screen Shot 2022-02-09 at 5 44 06 PM" src="https://user-images.githubusercontent.com/11668344/153303062-1b4217dc-b3ab-4c52-aaab-c64cf3307b72.png">

**VS Code extension or command-line**

Vscode: Version: 1.64.1 (Universal)
Python extension: v2022.0.1814523869
Pylance: v2022.2.0

**Comment by erictraut** [comment]
Thanks for the bug report. This will be fixed in the next release.

**Comment by erictraut** [comment]
This is included in pyright 1.1.221, which I just published. It will also be included in the next release of pylance.
